cleanlab/filter.py

```diff
@@ -902,41 +895,31 @@ def _prune_by_class(k: int, args=None) -> np.ndarray:
     k : int (between 0 and num classes - 1)
       The class of interest."""
 
-    if args:  # Single processing - params are passed in
-        (
-            labels,
-            label_counts,
-            prune_count_matrix,
-            pred_probs,
-            multi_label,
-            min_examples_per_class,
-        ) = args
-    else:  # Multiprocessing - data is shared across sub-processes
-        (
-            labels,
-            label_counts,
-            prune_count_matrix,
-            pred_probs,
-            multi_label,
-            min_examples_per_class,
-        ) = _get_shared_data()
+    k, min_examples_per_class, arrays = args
+    if arrays is None:
+        pred_probs = pred_probs_by_class[k]
+        prune_count_matrix = prune_count_matrix_cols[k]
+    else:
+        pred_probs = arrays[0]
+        prune_count_matrix = arrays[1]
 
-    if label_counts[k] > min_examples_per_class:  # No prune if not at least min_examples_per_class
-        num_issues = label_counts[k] - prune_count_matrix[k][k]
+    label_counts = pred_probs.shape[0]
+    label_issues_mask = np.zeros(label_counts, dtype=bool)
+    if label_counts > min_examples_per_class:  # No prune if not at least min_examples_per_class
+        num_issues = label_counts - prune_count_matrix[k]
         # Get return_indices_ranked_by of the smallest prob of class k for examples with noisy label k
-        label_filter = np.array([k in lst for lst in labels]) if multi_label else labels == k
         class_probs = pred_probs[:, k]
-        rank = np.partition(class_probs[label_filter], num_issues)[num_issues]
-        return label_filter & (class_probs < rank)
+        rank = np.partition(class_probs, num_issues)[num_issues]
+        return class_probs < rank
     else:
         warnings.warn(
             f"May not flag all label issues in class: {k}, it has too few examples (see argument: `min_examples_per_class`)"
         )
-        return np.zeros(len(labels), dtype=bool)
+        return np.zeros(label_counts, dtype=bool)
```

Comment:
try to extend existing unit test to cover this line of code as well, thanks!

Reply:
prune_by_class seems to have some bugs, will fix

Reply:
i realized that the previous code has another bug, but this one should rarely matter for real datasets: the prune_by_noise_rate and prune_by_count use np.partition to get a threshold, and then prunes with probs < threshold. This incurs bugs when there are more than one example that have probs exactly equal to the threshold value.

I changed the new code to using np.sort, which avoids this bug and is only very marginally slower, and the time difference is trivial compared to other time usages. Still doing other tests to make sure the outputs are correct

Reply:
@cgnorthcutt just FYI in case you want to have second look

Reply:
I put some logs about this in the notion doc. Some differences between the old and new versions do seem to be caused by this change, otherwise the new version seems correct from the test cases.

Reply:
I suspect so too, but @cgnorthcutt  would know forsure

Reply:
@clu0 is right. this is the correct fix.

I actually just discovered this same issue in my PR last week @clu0 @jwmueller  here:  https://github.com/cleanlab/cleanlab/blob/febcaa9375e0205c6f52c48c1295b02a348b9280/cleanlab/filter.py#L315

basically anything that depends on pred probs downstream can have this issue so I've also been switching my code to sort instead of using the O(n) algorithm.

looks good @clu0 !